File: android/mekuri/src/main/kotlin/app/mekuri/MekuriFoldedLeaf.kt
package app.mekuri

import android.graphics.RuntimeShader
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.runtime.Composable
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer

/**
 * A turning leaf with a different page printed on each side. [front] is the page
 * being read and [back] the page on its reverse, given in reading orientation.
 * Without a [back] the reverse is the front mirrored over one page. With one the
 * leaf spans a two-slot spread: its front fills the trailing slot, hinges at the
 * centre and lands in the leading slot. The leaf draws no shadow of its own; the
 * part of it short of the crease draws the departing page, opaque and undimmed.
 */
@Composable
internal fun MekuriFoldedLeaf(
    progress: () -> Float,
    direction: MekuriDirection,
    configuration: MekuriConfiguration,
    modifier: Modifier = Modifier,
    back: (@Composable () -> Unit)? = null,
    front: @Composable () -> Unit,
) {
    if (back == null) {
        MekuriFoldedPage(progress, direction, configuration, modifier, front)
        return
    }
    val isFolded by remember(progress) { derivedStateOf { progress() > 0f } }
    val mirrorScale = MekuriFoldPass.mirrorScale(direction)
    val frontShader = rememberMekuriFoldShader()
    val backShader = rememberMekuriFoldShader()
    Box(modifier) {
        if (!isFolded) {
            MekuriLeafSlot(mirrorScale, isMirroredInSlot = false, content = front)
            return@Box
        }
        MekuriLeafFace(frontShader, mirrorScale, MekuriFace.Front, configuration, direction, progress) {
            MekuriLeafSlot(mirrorScale, isMirroredInSlot = false, content = front)
        }
        MekuriLeafFace(backShader, mirrorScale, MekuriFace.Back, configuration, direction, progress) {
            MekuriLeafSlot(mirrorScale, isMirroredInSlot = true, content = back)
        }
    }
}

/**
 * The faces never overlap on screen, so the stacking order carries no meaning.
 */
@Composable
private fun MekuriLeafFace(
    shader: RuntimeShader,
    mirrorScale: Float,
    face: MekuriFace,
    configuration: MekuriConfiguration,
    direction: MekuriDirection,
    progress: () -> Float,
    content: @Composable () -> Unit,
) {
    Box(
        Modifier
            .fillMaxSize()
            .mekuriFold(
                shader = shader,
                mirrorScale = mirrorScale,
                face = face,
                isHinged = true,
                configuration = configuration,
                progress = { MekuriFoldPass(direction, progress()).shaderProgress },
            ),
    ) {
        content()
    }
}

/**
 * Placement is a geometry transform, never layout alignment: alignment follows
 * the layout direction and the leaf must follow the reading direction. The back
 * is mirrored within its own slot so both faces sample the trailing slot.
 */
@Composable
private fun MekuriLeafSlot(
    mirrorScale: Float,
    isMirroredInSlot: Boolean,
    content: @Composable () -> Unit,
) {
    Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        Box(
            Modifier
                .fillMaxWidth(SLOT_FRACTION)
                .fillMaxHeight()
                .graphicsLayer {
                    translationX = mirrorScale * size.width / 2f
                    scaleX = if (isMirroredInSlot) -1f else 1f
                },
        ) {
            content()
        }
    }
}

/** A leaf spans two slots and hinges at its centre. */
private const val SLOT_FRACTION = 0.5f

File: ios/Sources/Mekuri/MekuriFoldedLeaf.swift
import SwiftUI

/// A turning leaf with a different page printed on each side. `front` is the
/// page being read and `back` the page on its reverse, given in reading
/// orientation; the leaf mirrors it. Without a `back`, the reverse is the
/// front mirrored, exactly as ``MekuriFoldedPage`` draws it, over one page.
/// With one, the leaf spans a two-slot spread: its front fills the trailing
/// slot, hinges at the centre and lands in the leading slot. The leaf draws
/// no shadow; ``MekuriLeafShadow`` draws them beneath it.
struct MekuriFoldedLeaf<Front: View, Back: View>: View {
    private let progress: CGFloat
    private let direction: MekuriDirection
    private let configuration: MekuriConfiguration
    private let front: () -> Front
    private let back: (() -> Back)?

    init(
        progress: CGFloat,
        direction: MekuriDirection,
        configuration: MekuriConfiguration,
        @ViewBuilder front: @escaping () -> Front,
        @ViewBuilder back: @escaping () -> Back
    ) {
        self.progress = progress
        self.direction = direction
        self.configuration = configuration
        self.front = front
        self.back = back
    }

    init(
        progress: CGFloat,
        direction: MekuriDirection,
        configuration: MekuriConfiguration,
        @ViewBuilder front: @escaping () -> Front
    ) where Back == EmptyView {
        self.progress = progress
        self.direction = direction
        self.configuration = configuration
        self.front = front
        self.back = nil
    }

    var hasDistinctBackFace: Bool { self.back != nil }

    var isFolded: Bool { self.progress > 0 }

    var foldPass: MekuriFoldPass {
        MekuriFoldPass(direction: self.direction, progress: self.progress)
    }

    var body: some View {
        if let back = self.back {
            GeometryReader { proxy in
                if self.isFolded {
                    self.twoPassLeaf(size: proxy.size, back: back)
                } else {
                    self.inTrailingSlot(self.front(), size: proxy.size)
                }
            }
        } else {
            MekuriFoldedPage(progress: self.progress, direction: self.direction, configuration: self.configuration) {
                self.front()
            }
        }
    }

    /// The faces never overlap on screen, so the stacking order carries no
    /// meaning. Each pass is flattened before its shader runs so its
    /// translucent output composites exactly once. The back is mirrored
    /// within its slot so both faces sample the trailing slot.
    @ViewBuilder
    private func twoPassLeaf(size: CGSize, back: @escaping () -> Back) -> some View {
        let pass = self.foldPass
        ZStack {
            self.facePass(self.inTrailingSlot(self.front(), size: size), size: size, pass: pass, face: .front)
            self.facePass(self.inTrailingSlot(self.mirrored(back(), size: size), size: size), size: size, pass: pass, face: .back)
        }
        .frame(width: size.width, height: size.height)
    }

    private func facePass(_ layer: some View, size: CGSize, pass: MekuriFoldPass, face: MekuriFace) -> some View {
        layer
            .scaleEffect(x: pass.mirrorScale, y: 1)
            .compositingGroup()
            .layerEffect(self.fold(size: size, pass: pass, face: face), maxSampleOffset: size)
            .scaleEffect(x: pass.mirrorScale, y: 1)
    }

    /// Geometry effects only: layout alignment would follow the layout
    /// direction rather than the reading direction.
    private func inTrailingSlot(_ page: some View, size: CGSize) -> some View {
        let slotWidth = MekuriFoldShader.hinge(in: size)
        return page
            .frame(width: slotWidth, height: size.height)
            .offset(x: self.foldPass.mirrorScale * slotWidth / 2)
            .frame(width: size.width, height: size.height)
    }

    private func mirrored(_ page: some View, size: CGSize) -> some View {
        page
            .frame(width: MekuriFoldShader.hinge(in: size), height: size.height)
            .scaleEffect(x: -1, y: 1)
    }

    private func fold(size: CGSize, pass: MekuriFoldPass, face: MekuriFace) -> Shader {
        MekuriFoldShader.make(
            size: size,
            progress: pass.shaderProgress,
            configuration: self.configuration,
            face: face,
            hinge: MekuriFoldShader.hinge(in: size)
        )
    }
}
